### ISSUE TITLE
Regression from 1.44 to 1.45

### ISSUE THREAD

**Issue by mcarton** [rust-lang/rust#74739]
The following MVE prints `42` with [rustc 1.45.0](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:rust,selection:(endColumn:2,endLineNumber:11,positionColumn:1,positionLineNumber:1,selectionStartColumn:2,selectionStartLineNumber:11,startColumn:1,startLineNumber:1),source:'struct+Foo+%7B%0A++++x:+i32,%0A%7D%0A%0Afn+main()+%7B%0A++++let+mut+foo+%3D+Foo+%7B+x:+42+%7D%3B%0A++++let+x+%3D+%26mut+foo.x%3B%0A++++*x+%3D+13%3B%0A++++let+y+%3D+foo%3B%0A++++println!!(%22%7B%7D%22,+y.x)%3B+//+-%3E+42%3B+expected+result:+13%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1450,filters:(b:'0',binary:'0',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:rust,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'rustc+1.45.0+(Editor+%231,+Compiler+%231)+Rust',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,wrap:'1'),l:'5',n:'0',o:'%231+with+rustc+1.45.0',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) but `13` with [rustc 1.44.0](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:rust,selection:(endColumn:2,endLineNumber:11,positionColumn:1,positionLineNumber:1,selectionStartColumn:2,selectionStartLineNumber:11,startColumn:1,startLineNumber:1),source:'struct+Foo+%7B%0A++++x:+i32,%0A%7D%0A%0Afn+main()+%7B%0A++++let+mut+foo+%3D+Foo+%7B+x:+42+%7D%3B%0A++++let+x+%3D+%26mut+foo.x%3B%0A++++*x+%3D+13%3B%0A++++let+y+%3D+foo%3B%0A++++println!!(%22%7B%7D%22,+y.x)%3B+//+-%3E+42%3B+expected+result:+13%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1440,filters:(b:'0',binary:'0',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:rust,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'rustc+1.44.0+(Editor+%231,+Compiler+%231)+Rust',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,wrap:'1'),l:'5',n:'0',o:'%231+with+rustc+1.44.0',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

```rust
struct Foo {
    x: i32,
}

fn main() {
    let mut foo = Foo { x: 42 };
    let x = &mut foo.x;
    *x = 13;
    let y = foo;
    println!("{}", y.x); // -> 42; expected result: 13
}
```

Initially reported on [Stack Overflow](https://stackoverflow.com/questions/63087217/changes-using-mutable-reference-of-a-field-are-not-reflected-after-move-of-the-o).

**Comment by mcarton** [comment]
With beta or nightly this again prints 13 as expected.

**Comment by omrerdn** [comment]
```rust
struct Foo {
    x: i32,
}

fn main() {
    let mut foo = Foo { x: 42 };

    let x = &mut foo.x;
    *x = 13;
    let y = foo;
    println!("{:?}", (&y).x);  //only added this line
    println!("{:?}", y.x); //13
}
``` 
If there is any benefit, it works expected with this code.

**Comment by lcnr** [comment]
This is a duplicate of the already fixed #73609

It seems like this bug somehow slipped into beta :/

**Comment by SimonSapin** [comment]
@rust-lang/release Should there be a 1.45.1 with https://github.com/rust-lang/rust/issues/73609?

**Comment by peterjoel** [comment]
@SimonSapin Can 1.45.0 be yanked?

**Comment by jonas-schievink** [comment]
Yanking a release is not a thing. Presumably it wouldn't do anything once the stable channel is updated anyways.

**Comment by peterjoel** [comment]
@jonas-schievink It would do something for anyone who typed `rustup install 1.45.0`.

**Comment by benschulz** [comment]
I'm confused. Based on the examples this bug appears to be very impactful, i.e. it should break a lot of code. Yet it made it to stable which, I assume, involves a crater run and independent testing on beta. Is there a subtle condition to this bug that isn't met in most productive code?

**Comment by oli-obk** [comment]
The bug is almost impossible to trigger on real world code. You need all values that are going into the bug to be constant values and there can't be any control flow or function calls in between.

**Comment by paulocsanz** [comment]
Welp now we got a nice new test case. But we need a 1.45.1 release and 1.45.0 should (ideally) be yanked, although I don't know if the tooling gets crazy when we do it.

Has there ever been a rust version yanked? Because if not, it seems like a nice moment to have a discussion around making it better in the future. It won't be the last time we discover a huge problem that's an edge case (and we can fix it easily).

**Comment by wesleywiser** [comment]
As far as I know, `rustup` has no support for yanking releases. Doing so would probably have to be done manually by the infra team and it could cause more harm than good. For example, are current or old versions of `rustup` going to gracefully handle the case where a user has an installed release that doesn't exist anymore?  We've had other point releases for major issues in the past and users are generally quick to upgrade. 

There is already a `1.45.1` release scheduled for later this week with other backports and this fix will almost certainly be included.

**Comment by jieyouxu** [comment]
Could `rustup` try to print a massive warning (anytime it is invoked) telling the user that the patch is critical and is strongly recommended and that the previous release is deprecated?

**Comment by paulocsanz** [comment]
Yanking does not mean that the release does not exist tho, it means it can't be installed, that may generate issues if someone wants to install 1.45.0 to reproduce something for example.

But besides that, having a release tagged as no-no at least so the user realizes the problem and forces the install or installs the next minor version available.

Anyway, I think we should have a dynamic check that rustup/cargo makes from time to time (maybe when installing libs so you already are connected to the internet - it still doesn't seem clear the best moment to do that check tho) to know if the rustc version should be updated. Like we (should) do with CVEs.

This should rarely happen, but this seems like an appropriate time to integrate that feature (or just put it in the eventually pipeline).

**Comment by LeSeulArtichaut** [comment]
Assigning `P-critical` [as discussed as part of the Prioritization WG Procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2374739.20Regression.20from.201.2E44.20to.201.2E45).

**Comment by Mark-Simulacrum** [comment]
I am happy to include 5fa8b0880825d83eb01151e43e7de1e94e05cd2d in 1.45.1. @oli-obk -- there's some merge conflicts when cherry-picking onto stable, could you provide a commit/PR against stable with the merge conflicts fixed?

FWIW, I find the notion of putting warnings or saying "don't use this release" to be something we probably don't want to do right now. We'd basically have to do so for all non-latest Rust versions; they're all "unsupported" and may contain this (or other) problems. We ourselves do not patch more than the latest stable release if CVEs are detected, generally speaking. I would also ask that further discussion about yanking releases and such be directed to an internals thread and not this issue to avoid creating noise while we resolve this bug.

**Comment by oli-obk** [comment]
on it

**Comment by cysant** [comment]
Instead of yanking releases, wouldn't it be simpler to just not allow specifying patch versions when installing Rust? Instead of `rustup install 1.45.0` one would just `rustup install 1.45` and the latest patch version would be automatically installed. Since the Rust community uses Semantic Versioning, I don't see how that could be a problem. I know that theory and reality are not the same, but we have to assume the things we have stablished work as intended in order to make decisions.

The same could apply to crate versions.

**Comment by zhuxiujia** [comment]
Why this silly mistake，The production environment is going to break

**Comment by llacroix** [comment]
What @jieyouxu said.
It could be done with rustup check that will check and warn the user that new release introduce critical bug fixes or even security fixes.

**Comment by oli-obk** [comment]
@llacroix I don't think this is the place to discuss policy changes. This issue is specifically for fixing the regression, not how regression fixes are communicated in general.

**Comment by jonas-schievink** [comment]
This thread has served its purpose, so locking it.

**Comment by pnkfelix** [comment]
@lcnr [commented](https://github.com/rust-lang/rust/issues/74739#issuecomment-663848055): 

> This is a duplicate of the already fixed #73609

by the way, this issue was not, technically, a duplicate of #73609.

Or at least, the issue described in #73609 only surfaces under conditions witnessed after commit 38bd83df88288f2f8d1fc2dd317189cac3825920

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.5.2</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc 2020-04-01 --end 2020-06-22 -- run
```
</details>

while this issue surfaces under conditions witnessed after commit 3fe4dd2dda2826643c4ce4ee7307707a90e08d25

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.5.2</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc 2020-05-01 --end 2020-06-01 -- run
```
</details>

It does seem like the same patch fixes both issues; i.e., it seems likely that there is indeed *one bug* underlying both distinct issues.

But the fact that they are distinct issues helps explain why we did not backport the fix to beta way back when.
